### ISSUE TITLE
NullPointerException in Release 0.7.1

### ISSUE THREAD

**Issue by adoelger** [Azure/azure-event-hubs#168]
Hi,

I'm using the latest (0.7.1) event hub java sdk and get sporadically the follwoing NullPointerException:  

Jun 14, 2016 2:10:39 PM com.microsoft.azure.servicebus.MessagingFactory$RunReactor run
WARNING: UnHandled exception while processing events in reactor:
java.lang.NullPointerException
org.apache.qpid.proton.engine.impl.EventImpl.dispatch(EventImpl.java:112)
org.apache.qpid.proton.reactor.impl.ReactorImpl.dispatch(ReactorImpl.java:304)
org.apache.qpid.proton.reactor.impl.ReactorImpl.process(ReactorImpl.java:273)
com.microsoft.azure.servicebus.MessagingFactory$RunReactor.run(MessagingFactory.java:341)
java.lang.Thread.run(Thread.java:745)Cause: null
org.apache.qpid.proton.reactor.impl.IOHandler.onUnhandled(IOHandler.java:303)
org.apache.qpid.proton.engine.BaseHandler.handle(BaseHandler.java:236)
org.apache.qpid.proton.engine.impl.EventImpl.dispatch(EventImpl.java:108)
org.apache.qpid.proton.reactor.impl.ReactorImpl.dispatch(ReactorImpl.java:304)
org.apache.qpid.proton.reactor.impl.ReactorImpl.process(ReactorImpl.java:273)
com.microsoft.azure.servicebus.MessagingFactory$RunReactor.run(MessagingFactory.java:341)
java.lang.Thread.run(Thread.java:745)

**Comment by hmlam** [comment]
are you using PartitionReceiver.receive() or the PartitionReceiver.setReceiveHandler() API? as part of EPH work we do have a fix in the setReceiveHandler() code path, so I want to see if this is an known issue or not.

**Comment by adoelger** [comment]
No, we are doing a synchronous send!

**Comment by wofr** [comment]
Hi !
I do have the same problem with 0.7.2. 
I'm using the asyncSend. The exception ending up on system.out as they are not catched by my exception handling because the thrown in the background. In parallel with this exception the write performance starts to drop dramatically. I use severall clients in paralle and it seems if this exception occurs the client is some how dead. My problem is I've no way to catch this exception as it is not in my context.

java.lang.Thread.run(Unknown Source)Cause: null
Jun 30, 2016 4:08:57 PM com.microsoft.azure.servicebus.MessagingFactory$RunReactor run
WARNING: UnHandled exception while processing events in reactor:
java.lang.NullPointerException
org.apache.qpid.proton.engine.impl.EventImpl.dispatch(EventImpl.java:112)
org.apache.qpid.proton.reactor.impl.ReactorImpl.dispatch(ReactorImpl.java:307)
org.apache.qpid.proton.reactor.impl.ReactorImpl.process(ReactorImpl.java:276)
com.microsoft.azure.servicebus.MessagingFactory$RunReactor.run(MessagingFactory.java:341)

**Comment by SreeramGarlapati** [comment]
@wofr - we are working on a fix for this and expect to release an update by 7/25.

**Comment by hmlam** [comment]
https://github.com/Azure/azure-event-hubs/pull/187 should address this issue.

**Comment by SreeramGarlapati** [comment]
@wofr
Can you please try our release candidate for 0.7.4 & verify this issue:

<repository> 
<id>azure eventhubs</id> 
<url>https://oss.sonatype.org/content/repositories/commicrosoftazure-1333/</url> 
</repository>
<dependency>
<groupId>com.microsoft.azure</groupId>
<artifactId>azure-eventhubs</artifactId>
<version>0.7.4</version>
</dependency>

**Comment by wofr** [comment]
thanks! I will give it a try. But it seems I can not access the repo 
"404 - Repository "commicrosoftazure-1333 (staging: open)" [id=commicrosoftazure-1333] exists but is not exposed"

**Comment by SreeramGarlapati** [comment]
oops, my bad - fixed the issue. Please give it another try.

**Comment by wofr** [comment]
I'm impressed!! I was abel to send avg >4000 msgs per second (used 4 clients) no timouts, no exception nothing :). I kept it running for hours no problems at all. 
Great work!

**Comment by wofr** [comment]
….now I made some wired observation, but this time it is a read issue. I do have a reader connected reading from the end of the Event-Hub. Even so I know no new messages are in the Hub periodically my onReceive function is called with empty events (null)

`partitionReceiver.setReceiveHandler(new PartitionReceiveHandler(100) {`
`  override def onError(throwable: Throwable): Unit = {`
`    log.error("PartitionReceiver failed with error {}", throwable) `
`    self ! EventHubPartitionReceiverFailed(partitionId,throwable.getMessage) `
`  }`

`override def onReceive(events: java.lang.Iterable[EventData]): Unit = {`
`	log.info("Recieved events {}",events)   // log output “INFO  Recieved events null ” `
`      readFromStreamAnalyticsPreProcessedStream( events.asScala) `
`  }`

Should I create an issue for it? I switched form 0.6.9 to this version so I don’t know when this issue occurred the first time. I did not observe the problem with 0.6.9

**Comment by SreeramGarlapati** [comment]
Thanks a lot for Staying with us so Patiently @wofr. I truly appreciate that!

The behavior with the reader invoking the call back on null - is expected. We apologize for not documenting this. We had to add this behavior to support a Knob in EventProcessorHost (https://github.com/Azure/azure-event-hubs/blob/master/java/azure-eventhubs-eph/src/main/java/com/microsoft/azure/eventprocessorhost/EventProcessorOptions.java#L170).

**Comment by wofr** [comment]
You are wellcome! :) I really appreciate the improvments of the last 3 month and the latest version is really great.    

The null on the call back is some how unexpected. So adding this in the documentation is definitly a good idea.

**Comment by hmlam** [comment]
I believe the null of callback is something that we need to improve upon and Sreeram already knows what needs to be done. Please file a new issue to track that work, and let's close this one since https://github.com/Azure/azure-event-hubs/pull/187 has fixed it.
